src/algebra/big_operators/basic.lean

```diff
@@ -1649,8 +1649,69 @@ end list
 
 namespace multiset
 
+lemma disjoint_list_sum_left {a : multiset α} {l : list (multiset α)} :
+  multiset.disjoint l.sum a ↔ ∀ b ∈ l, multiset.disjoint b a :=
+begin
+  induction l with b bs ih,
+  { simp only [zero_disjoint, list.not_mem_nil, is_empty.forall_iff, forall_const, list.sum_nil], },
+  { simp_rw [list.sum_cons, disjoint_add_left, list.mem_cons_iff, forall_eq_or_imp],
+    simp [and.congr_left_iff, iff_self, ih], },
+end
+
+lemma disjoint_list_sum_right {a : multiset α} {l : list (multiset α)} :
+  multiset.disjoint a l.sum ↔ ∀ b ∈ l, multiset.disjoint a b :=
+by simpa only [disjoint_comm] using disjoint_list_sum_left
+
+lemma disjoint_sum_left {a : multiset α} {i : multiset (multiset α)} :
+  multiset.disjoint i.sum a ↔ ∀ b ∈ i, multiset.disjoint b a :=
+quotient.induction_on i $ λ l, begin
+  rw [quot_mk_to_coe, multiset.coe_sum],
+  exact disjoint_list_sum_left,
+end
+
+lemma disjoint_sum_right {a : multiset α} {i : multiset (multiset α)} :
+  multiset.disjoint a i.sum ↔ ∀ b ∈ i, multiset.disjoint a b :=
+by simpa only [disjoint_comm] using disjoint_sum_left
+
+lemma disjoint_finset_sum_left {β : Type*} {i : finset β} {f : β → multiset α} {a : multiset α} :
+  multiset.disjoint (i.sum f) a ↔ ∀ b ∈ i, multiset.disjoint (f b) a :=
+begin
+  convert (@disjoint_sum_left _ a) (map f i.val),
+  simp [finset.mem_def, and.congr_left_iff, iff_self],
+end
+
+lemma disjoint_finset_sum_right {β : Type*} {i : finset β} {f : β → multiset α} {a : multiset α} :
+  multiset.disjoint a (i.sum f) ↔ ∀ b ∈ i, multiset.disjoint a (f b) :=
+by simpa only [disjoint_comm] using disjoint_finset_sum_left
+
 variables [decidable_eq α]
 
+lemma disjoint_finset_iff_sum_eq_sup {β : Type*} {i : finset β} {f : β → multiset α} :
+  i.sum f = i.sup f ↔ ∀ x y ∈ i, x ≠ y → multiset.disjoint (f x) (f y) :=
+begin
+  induction i using finset.cons_induction_on with z i hz hr,
+  { simp only [finset.not_mem_empty, is_empty.forall_iff, implies_true_iff,
+      finset.sum_empty, finset.sup_empty, bot_eq_zero, eq_self_iff_true], },
+  { simp_rw [finset.sum_cons hz, finset.sup_cons, finset.mem_cons, multiset.sup_eq_union],
+    split,
```

Comment:
Yes, I tried to shorten it but was not very successful. I'll try some more but feel free to improve it.

Reply:
Well, I was still not really able to make this proof shorter. The crux of the proof appears to prove the following statement: 
```f z + i.sum f = f z ∪ i.sup f ↔ (i.sum f = i.sup f) ∧ (f z).disjoint (i.sum f)``` 
(although this statement does not appear explicitly in the proof). I thought about making this statement into a general lemma and then use it to shorten the proof: 
```lean
example {x y z : multiset α} (h : y ≤ x) : z + x = z ∪ y ↔ x = y ∧ z.disjoint x
```
but I am not sure this lemma would have much use.

Reply:
I just had a go, was able to reduce the proof to the same point

<details>

```lean
lemma finset_sum_eq_sup_iff_disjoint {β : Type*} {i : finset β} {f : β → multiset α} :
  i.sum f = i.sup f ↔ ∀ x y ∈ i, x ≠ y → multiset.disjoint (f x) (f y) :=
begin
  induction i using finset.cons_induction_on with z i hz hr,
  { simp only [finset.not_mem_empty, is_empty.forall_iff, implies_true_iff,
      finset.sum_empty, finset.sup_empty, bot_eq_zero, eq_self_iff_true], },
  { simp_rw [finset.sum_cons hz, finset.sup_cons, finset.mem_cons, multiset.sup_eq_union,
      forall_eq_or_imp, ne.def, eq_self_iff_true, not_true, is_empty.forall_iff, true_and,
      imp_and_distrib, forall_and_distrib, ←hr, @eq_comm _ z],
    have := λ x ∈ i, ne_of_mem_of_not_mem H hz,
    simp only [this, not_false_iff, true_implies_iff] {contextual := tt},
    simp_rw [←disjoint_finset_sum_left, ←disjoint_finset_sum_right, disjoint_comm, ← and_assoc,
      and_self],
    sorry }
end
```

</details>

Your lemma suggestion looks like a great idea to me, although I'd adapt it slightly to make the orientation of the `and` match the order of the arguments:
```lean
example {x y z : multiset α} (h : z ≤ y) : x + y = x ∪ z ↔ y = z ∧ x.disjoint y
example {x y z : multiset α} (h : y ≤ x) : z + x = z ∪ y ↔ z.disjoint x ∧ x = y
```

Reply:
I added the two lemmas and the proof of ```finset_sum_eq_sup_iff_disjoint``` is just one extra line to your code now :) 

The proof of ```add_eq_union_left_of_le``` is very crude and needs to be refined but I have to go to bed now. You are welcome to improve on it if you have time otherwise I should be able to find some time tomorrow to improve it.